### PR TITLE
[ci][7up/3] initial dependency upgrades

### PR DESCRIPTION
**Pull request by can-anyscale**
Upgrade ray dependencies to versions that are available on both python 3.9 and python 3.11. ray 3.8 is no longer build-able with these new versions.

All tests passed except for two ml tests which the team has agreed to deprecate them w00t w00t.

Test:
- CI

### CODE REVIEW

File: .buildkite/build.rayci.yml [hunk 1]
```diff
@@ -158,5 +152,7 @@ steps:
       - ./ci/ci.sh compile_pip_dependencies
       - cp -f ./python/requirements_compiled.txt /artifact-mount/
     soft_fail: true
+    job_env: oss-ci-base_build
```
Comment (can-anyscale) [inline .buildkite/build.rayci.yml:155]: forge is python 3.8 so need to use this image that is python 3.9

File: ci/ci.sh [hunk 2]
```diff
@@ -112,7 +112,7 @@ compile_pip_dependencies() {
   python -c "import torch" 2>/dev/null && HAS_TORCH=1
   pip install --no-cache-dir numpy torch
 
-  pip-compile --resolver=backtracking -q \
+  pip-compile --verbose --resolver=backtracking \
```
Comment (can-anyscale) [inline ci/ci.sh:115]: without `verbose` it's impossible to resolve conflicts

File: ci/docker/ml.build.Dockerfile [hunk 3]
```diff
@@ -24,7 +24,7 @@ set -euo pipefail
 # TODO (can): Move mosaicml to train-test-requirements.txt
 pip install "mosaicml==0.12.1"
 DOC_TESTING=1 TRAIN_TESTING=1 TUNE_TESTING=1 DATA_PROCESSING_TESTING=1 \
-  INSTALL_HOROVOD=1 INSTALL_TIMESERIES_LIBS=1 INSTALL_HDFS=1 \
```
Comment (can-anyscale) [inline ci/docker/ml.build.Dockerfile:27]: timeseries libs are moved into train-requirements.txt

File: ci/ray_ci/ml.tests.yml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 flaky_tests:
   # non-gpu tests
   - //python/ray/train:horovod_cifar_pbt_example
-  # doc tests
-  # gpu tests
+  # rely on scikit-optimize that cannot be upgraded
+  - //python/ray/tune:test_sample
```
Comment (can-anyscale) [inline ci/ray_ci/ml.tests.yml:5]: ml team will deprecate these tests later on

File: python/ray/autoscaler/_private/aws/node_provider.py [hunk 5]
```diff
@@ -431,22 +431,12 @@ def _create_node(self, node_config, tags, count):
...
-                        # {"code": "0", "Message": "pending"}
-                        state_reason = instance.state_reason or {"Message": "pending"}
-
                         cli_logger.print(
                             "Launched instance {}",
                             instance.instance_id,
                             _tags=dict(
                                 state=instance.state["Name"],
-                                info=state_reason["Message"],
+                                info=instance.state_reason["Message"] or "pending",
```
Comment (can-anyscale) [inline python/ray/autoscaler/_private/aws/node_provider.py]: fix for new value in python 3.9

File: python/ray/autoscaler/_private/aws/node_provider.py [hunk 6]
```diff
@@ -439,14 +439,18 @@ def _create_node(self, node_config, tags, count):
 
                         # The correct value is technically
                         # {"code": "0", "Message": "pending"}
-                        state_reason = instance.state_reason or {"Message": "pending"}
+                        state_reason = "pending"
+                        if instance.state_reason:
+                            state_reason = (
+                                instance.state_reason["Message"] or state_reason
```
Comment (can-anyscale) [inline python/ray/autoscaler/_private/aws/node_provider.py:445]: in python 3.9, even when instance.state_reason is not None, it's 'Message' entry can still be None

**Comment by can-anyscale** [comment]
Putting up for early feebdack on the package versions I'm using

**Review by architkulkarni — APPROVED** [review]
node provider change looks good!

File: ci/docker/data.build.Dockerfile [hunk 7]
```diff
@@ -20,8 +20,14 @@ COPY . .
 RUN <<EOF
 #!/bin/bash
 
+set -ex
+
 DATA_PROCESSING_TESTING=1 ARROW_VERSION=$ARROW_VERSION \
   ARROW_MONGO_VERSION=$ARROW_MONGO_VERSION ./ci/env/install-dependencies.sh
+if [[ -n $ARROW_MONGO_VERSION ]]; then
```
Comment (aslonnie) [inline ci/docker/data.build.Dockerfile:27]: double quote `"$ARROW_MONGO_VERSION"`?

File: python/requirements/ml/tune-requirements.txt [hunk 8]
```diff
@@ -7,9 +7,7 @@ bayesian-optimization==1.4.3
 ConfigSpace==0.7.1
 hpbandster==0.7.4
 
-# Todo: Use version pin again after HEBO is updated on pypi
-git+https://github.com/huawei-noah/HEBO@9a2a674c22518eed35a8b98e5134576741a95410#subdirectory=HEBO
```
Comment (aslonnie) [inline python/requirements/ml/tune-requirements.txt:11]: where is `huawei-noah` going? we do not need it anymore?

File: python/requirements/ml/tune-test-requirements.txt [hunk 9]
```diff
@@ -2,16 +2,21 @@
 
 aim==3.17.5
 
-gpy==1.10.0
+gpy==1.10.0; python_version <= '3.9'
```
Comment (aslonnie) [inline python/requirements/ml/tune-test-requirements.txt:5]: what version of `gpy` will be used for other python versions?

**Review by aslonnie — COMMENTED** [review]
I am in general a bit worried about the additional constraints that are set for the additional python packages, but I do not know what are really the best options to move forward.

**Review by aslonnie — APPROVED** [review]
the build file changes look okay; just to unblock.

**Review by jjyao — APPROVED** [review]

File: ci/ray_ci/ml.tests.yml (hunk 4 above)
Comment (matthewdeng) [reply ci/ray_ci/ml.tests.yml:5]: Let me put up a PR to remove `scikit-optimize` now.

File: python/requirements/ml/tune-requirements.txt [hunk 10]
```diff
@@ -7,9 +7,7 @@ bayesian-optimization==1.4.3
...
-# HEBO==0.3.4
+HEBO==0.3.5; python_version <= '3.9'
```
Comment (matthewdeng) [inline python/requirements/ml/tune-requirements.txt:10]: Any reason for the python upper bound?

Comment (can-anyscale) [reply python/requirements/ml/tune-requirements.txt:10]: ah good catch, I'll remove the bound

File: python/requirements/ml/tune-requirements.txt (hunk 8 above)
Comment (can-anyscale) [reply python/requirements/ml/tune-requirements.txt:11]: we can use the pypi version now

File: python/requirements/ml/tune-test-requirements.txt (hunk 9 above)
Comment (can-anyscale) [reply python/requirements/ml/tune-test-requirements.txt:5]: the latest version which is 0.3.5 will break so I'll upgrade it for python 3.11 in another PR

File: ci/ray_ci/ml.tests.yml (hunk 4 above)
Comment (matthewdeng) [reply ci/ray_ci/ml.tests.yml:5]: PR here: https://github.com/ray-project/ray/pull/43969

Comment (can-anyscale) [reply ci/ray_ci/ml.tests.yml:5]: love it!

File: python/requirements/ml/data-requirements.txt [hunk 11]
```diff
@@ -6,6 +6,7 @@ aioboto3==11.2.0
 crc32c==2.3
 flask_cors
 modin==0.22.2
-raydp>=0.0.dev0
+pandas==1.5.3
```
Comment (scottjlee) [inline python/requirements/ml/data-requirements.txt:9]: is the pandas pin required for data specifically?

File: python/requirements/ml/data-test-requirements.txt [hunk 12]
```diff
@@ -1,11 +1,13 @@
 # Used by CI for datasets tests.
 # https://github.com/ray-project/ray/pull/29448#discussion_r1006256498
 
-python-snappy
-tensorflow-datasets
+python-snappy; python_version <= '3.10'
```
Comment (scottjlee) [inline python/requirements/ml/data-test-requirements.txt:4]: is the python upper bound required here?

**Review by scottjlee — APPROVED** [review]

File: python/requirements/ml/data-requirements.txt (hunk 11 above)
Comment (can-anyscale) [reply python/requirements/ml/data-requirements.txt:9]: i was just helping out pip a little bit in its quest to resolve dependencies, not specific to data; 1.5.3 is essentially the latest 1.* version of pandas

File: python/requirements/ml/data-test-requirements.txt (hunk 12 above)
Comment (can-anyscale) [reply python/requirements/ml/data-test-requirements.txt:4]: good catch, I'll dig deeper in the subsequent PR where I add python 3.11 (and potentially need to fixes test failure fallout from that); i'll keep it pinned for <3.11 for now

**Review by matthewdeng — APPROVED** [review]
<div><br/></div>

**Pull request merged**